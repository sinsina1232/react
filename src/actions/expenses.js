import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch)=>{
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt }

    return database.ref('expenses').push(expense).then((ref)=>{
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {})=>{
  return (dispatch) =>{
  return database.ref(`expenses/${id}`).remove().then(()=>{
    dispatch(removeExpense({id}));
});
};
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});
export const startEditExpense = (id, updates) =>{
  return (dispatch)=>{
    return database.ref(`expenses/${id}`).update(updates).then(()=>{
      dispatch(editExpense(id, updates));
    });
  };
};

export const setExpenses = (expenses) =>({
  type: 'SET_EXPENSES',
  expenses
});

//seting the startin point

export const startSetExpenses = () => {
  return (dispatch) =>{
  return database.ref('expenses').once('value').then((snapshot)=>{
   const expensesData = [];
   snapshot.forEach((childSnapshot)=>{
     expensesData.push({
      id: childSnapshot.key,
      ...childSnapshot.val()
    });
  });
  dispatch(setExpenses(expensesData));
});

};
};


//1. fetch all the data once,
//2. parse the data into an array
//3. dispatch SET_EXPENSES






///
