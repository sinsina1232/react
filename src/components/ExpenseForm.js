import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

//const date = new Date();
const now = moment();
console.log(now.format('MMMM Do, YYYY'));
export default class ExpenseForm extends React.Component{
  state = {
    description:'',
    note:'',
    amount: '',
    createdAt: moment(),
    calendarFocused: false
  };
  onDescriptionChange = (e)=>{
    const description = e.target.value;
    this.setState(()=>({ description }));
  };
  onChangeTextArea = (e)=>{
    const note = e.target.value;
    this.setState(()=>({ noteue }));
  };
  onAmountChange = (e) =>{
    const amount = e.target.value;
    if(amount.match(/^\d{1,9}(\.\d{0,2})?$/)){
        this.setState(()=>({ amount }));
    }
  };
    onDateChange = (createdAt) =>{
      this.setState(()=>{ createdAt });
    };
    onFocusChange = ({ focused }) =>{
      this.setState(()=>({ calendarFocused: focused }));
    };

  render(){
    return (
      <div>
        <form>
            <input
                type="text"
                placeholder="description"
                value={this.state.desciption}
                onChange={this.onDescriptionChange}
                autoFocus
              />
            <input
                type="text"
                value={this.state.amount}
                placeholder="Amount"
                onChange={this.onAmountChange}
            />

            <SingleDatePicker
              date= {this.state.createdAt}
              onDateChange={this.onDateChange}
              focused= {this.state.calendarFocused}
              onFocusChange={this.onFocusChange}

            />
            <textarea
                placeholder="Add a note for your input*(optional)"
                value={this.state.note}
                onChange={this.onChangeTextArea}
            />
            <button>
              Add Expense
            </button>
        </form>
      </div>
    )
  }
}
