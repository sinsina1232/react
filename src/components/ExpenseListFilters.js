import React from 'react';
import { connect } from 'react-redux';
import SET_TEXT_FILTER from '../reducers/filters';

const ExpenseListFilters = (props) => (
  <div>
    <input type="text" value={props.filters.text} onChange={(e) => {
        props.dispatch(setTextFilter(e.target.value));
    }} />
  </div>
);
const mapStateToProps =(state) =>{
  return{
    filters: state.filters
  }

}
export default connect(mapStateToProps)(ExpenseListFilters);
