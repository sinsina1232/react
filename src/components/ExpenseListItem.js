// export stateless functional component
// ender desctiption, amount, createdAt
import { Link } from 'react-router-dom';
import React from 'react';

const ExpenseListItem = ({dispatch, id, description, amount, createdAt})=>(
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
      </Link>

    <p >{amount} - {createdAt}</p>

    </div>
  );
export default ExpenseListItem;
