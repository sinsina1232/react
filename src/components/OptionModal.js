import React from 'react';
import Modal from 'react-modal';
const OptionModal = (props)=>(
    <Modal
      isOpen= {!!props.selectedOption}
      contentLabel = "Selected Option"
      onRequestClose = {props.handleOkayOption}
      closeTimeoutMS={200}
      className="modal"
      >
    <h1 className="modal__title">Selected Option</h1>
    {props.selectedOption && <p className="modal__body">{props.selectedOption}</p>}
    <button className="button" onClick={props.handleOkayOption}>Okay</button>
    </Modal>
  );
export default OptionModal;
