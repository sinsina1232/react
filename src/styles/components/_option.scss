.option{
  border-bottom: 1px solid lighten($light-blue, 10%);
  display: flex;
  justify-content: space-between;
  padding: $l-size $m-size;
}
.option__text{
  font-size: 2rem;
  font-weight: 500;
  color: white;
  margin: 0;
  word-break: break-all;
}
