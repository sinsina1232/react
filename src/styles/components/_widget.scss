//widget header
.widget{
  background: $light-blue;
  margin-bottom: $xl-size;
}
//create  a widget elemetnt selector (message) -- use it on the p tag
// offwhite/ no margin- l-size padding all sides text-align:center border-buttom 1px solid lighted version of the light blue. llighten it 10; percenat
.widget-header{
  background: $blue;
  display: flex;
  color: $offwhite;
  justify-content: space-between;
  padding: $m-size;
}
.widget-header__title{
  margin: 0;
}
.widget__message{
  text-align: center;
  color: $offwhite;
  margin: 0;
  padding: $l-size;
  border-bottom: 1px solid lighten($light-blue, 10%);
}
