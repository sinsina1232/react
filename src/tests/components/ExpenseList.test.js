import { shallow } from 'enzyme';
import React from 'react';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

test('should render expense list from expenses',()=>{
  const wrapper = shallow(<ExpenseList expenses={expenses} /> );
  expect(wrapper).toMatchSnapshot();
});
test('should render the message on the empty array',()=>{
  const wrapper = shallow(<ExpenseList expenses={[]} /> );
  expect(wrapper).toMatchSnapshot();
});
