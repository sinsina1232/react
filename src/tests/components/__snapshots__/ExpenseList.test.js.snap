// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list from expenses 1`] = `
<div>
  <h1>
    Expense list 
  </h1>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Gum"
    id={1}
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={19500}
    createdAt={-345600000}
    description="Rent"
    id={2}
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={4500}
    createdAt={345600000}
    description="Credit cards"
    id={3}
    key="3"
    note=""
  />
</div>
`;

exports[`should render the message on the empty array 1`] = `
<div>
  <h1>
    Expense list 
  </h1>
  <p>
    Thre is no expense to show, you can add one now
  </p>
</div>
`;
